packages/cli/src/utils/hcms.ts

```diff
@@ -0,0 +1,182 @@
+  }>
+}
+
+export function splitCustomDefinitions(
+  coreDefinitions: ContentTypeOrSectionDefinition[],
+  customDefinitions: ContentTypeOrSectionDefinition[],
+  primaryIdentifier: 'id' | 'name'
+) {
+  const coreDefinitionIdentifiers = new Set<string>(
+    coreDefinitions.map((definition) => definition[primaryIdentifier] ?? '')
+  )
+
+  const duplicates: ContentTypeOrSectionDefinition[] = []
+  const newDefinitions: ContentTypeOrSectionDefinition[] = []
+
+  customDefinitions.forEach((definition) => {
+    if (!definition[primaryIdentifier]) {
+      console.error('Ignoring invalid definition:', definition)
+      return
+    }
+
+    if (
+      coreDefinitionIdentifiers.has(definition[primaryIdentifier] as string)
+    ) {
+      duplicates.push(definition)
+      return
+    }
+
+    newDefinitions.push(definition)
+  })
+
+  return { duplicates, newDefinitions }
+}
+
+export function dedupeAndMergeDefinitions(
+  coreDefinitions: ContentTypeOrSectionDefinition[],
+  duplicates: ContentTypeOrSectionDefinition[],
+  primaryIdentifier: 'id' | 'name'
+) {
+  const sortedCoreDefs = coreDefinitions.filter((definition) =>
+    Boolean(definition[primaryIdentifier])
+  )
+  sortedCoreDefs.sort((a, b) =>
+    (a[primaryIdentifier] as string) < (b[primaryIdentifier] as string) ? -1 : 1
+  )
+
+  const sortedDuplicates = duplicates.filter((definition) =>
+    Boolean(definition[primaryIdentifier])
+  )
+  sortedDuplicates.sort((a, b) =>
+    (a[primaryIdentifier] as string) < (b[primaryIdentifier] as string) ? -1 : 1
+  )
+
+  let duplicateIdx = 0
+
+  const result = sortedCoreDefs.map((currentDefinition) => {
+    const isDuplicateMatch =
+      currentDefinition[primaryIdentifier] ===
+      sortedDuplicates[duplicateIdx]?.[primaryIdentifier]
+
+    if (duplicateIdx < sortedDuplicates.length && isDuplicateMatch) {
+      return sortedDuplicates[duplicateIdx++]
+    }
+
+    return currentDefinition
+  })
+
+  return result
+}
+
+async function confirmUserChoice(
+  duplicates: ContentTypeOrSectionDefinition[],
+  fileName: string
+) {
+  const goAhead = await CliUx.ux.confirm(
+    `You are about to override default ${
+      fileName.split('.')[0]
+    }:\n\n${duplicates
+      .map((definition) => definition.id || definition.name)
+      .join('\n')}\n\nAre you sure? [yes/no]`
+  )
+
+  if (!goAhead) {
+    throw new Error('cms-sync cancelled by user.')
+  }
+
+  return
+}
+
+async function mergeCMSFile(fileName: string) {
+  const coreFilePath = path.join(coreCMSDir, fileName)
+  const customFilePath = path.join(userCMSDir, fileName)
+
+  const coreFile = readFileSync(coreFilePath, 'utf8')
+  const coreDefinitions: ContentTypeOrSectionDefinition[] = JSON.parse(coreFile)
+
+  const primaryIdentifierForDefinitions =
+    fileName === 'content-types.json' ? 'id' : 'name'
+
+  let output: ContentTypeOrSectionDefinition[] = []
+
+  // TODO: create a validation when has the cms files but doesn't have a component for then
```

Comment:
What do you think about this text?
```suggestion
  // TODO: create a validation when the CMS files exist but don't have a component for them
```

Reply:
Thanks! This was already there, and I was lazy to update 😬